pruning/manager.go

```diff
@@ -0,0 +1,216 @@
+package pruning
+
+import (
+	"container/list"
+	"encoding/binary"
+	"fmt"
+	"sync"
+
+	"github.com/cosmos/cosmos-sdk/pruning/types"
+
+	"github.com/tendermint/tendermint/libs/log"
+	dbm "github.com/tendermint/tm-db"
+)
+
+type Manager struct {
+	logger               log.Logger
+	opts                 *types.PruningOptions
+	snapshotInterval     uint64
```

Comment:
Do you think its better for pruning and snapshots to be managed separately, or should one have a reference to the other?

Reply:
Pruning can work independently from snapshots if those are not enabled. In that case, `snapshotInterval` is 0 and we do not skip any heights.

However, if snapshot settings are set, then the snapshot manager is also set and the `snapshotInterval` within the `pruningManager` is non-negative (it is set [here](https://github.com/osmosis-labs/cosmos-sdk/blob/8d4d6c7d718804d65330ff24a285392a74434131/pruning/manager.go#L113)).

In such a case, `snapshotInterval` functions as the old `pruning-keep-every` so that there is no race between pruning and snapshots where we get the "active reader" error in iavl. There is one important difference - when the snapshot is complete, `snapshot.Manager` "tells" `pruning.Manager` to prune that height  away with the following call:
https://github.com/osmosis-labs/cosmos-sdk/blob/8d4d6c7d718804d65330ff24a285392a74434131/snapshots/manager.go#L139

As a result, the `snapshotInterval` height is _eventually_ pruned.

I agree that pruning and snapshot managers are coupled to each other. However, pruning and snapshots are coupled functionally so there is no way without creating the 2 components without these interactions. Functionally, the 2 components are independent - you can set pruning settings without snapshots and you can set snapshot settings without pruning.

Please let me know if I can explain anything further or if you have any suggestions on what to change.

Reply:
To help with understanding this complexity, I can create UML and sequence diagrams if you would like me to

Reply:
Oh the pruning manager is being set here: https://github.com/osmosis-labs/cosmos-sdk/pull/140/files#diff-84a22ba89cc7868bbafc59e07c35ba0bffb05115962cb885bf39eef1a1a8a66eR70

So: Pruning manager always needed, Snapshot manager optional, if snapshot manager set, must update a certain param in the pruning manager? It does so on initialization of snapshot manager.

If this parameter is set, the pruning manager will not delete any snapshot height, and it is up to the snapshot manager to prune it?

I don't think we need a UML diagram, would be convenient if we can add a note to this effect in the README for the pruning and snapshot folder though. (Unless theres a better place for it to live) And on the SetSnapshotInterval function https://github.com/osmosis-labs/cosmos-sdk/pull/140/files#diff-309b9123bdfc80cc367cc95308580b5489e26406b9f5f0920ac55c5b198a5a33R112-R115

Reply:
Also if this is the case, the interaction front now 'clicks' for me!

Reply:
> Oh the pruning manager is being set here: https://github.com/osmosis-labs/cosmos-sdk/pull/140/files#diff-84a22ba89cc7868bbafc59e07c35ba0bffb05115962cb885bf39eef1a1a8a66eR70

Yes, that's how the pruning manager gets its `snapshotInterval` member field updated if the snapshot options/manager are set

> So: Pruning manager always needed, Snapshot manager optional, if snapshot manager set, must update a certain param in the pruning manager? It does so on initialization of snapshot manager.

That's correct

> If this parameter is set, the pruning manager will not delete any snapshot height, and it is up to the snapshot manager to prune it?

That is mostly correct with one subtle difference - snapshot manager "tells" pruning manager to prune the snapshot height once it is done by calling [this](https://github.com/osmosis-labs/cosmos-sdk/blob/8d4d6c7d718804d65330ff24a285392a74434131/snapshots/manager.go#L139) method. That is, the snapshot manager does not prune itself. Instead, it delegates that to the pruning manager.

> would be convenient if we can add a note to this effect in the README for the pruning and snapshot folder though.

Will work on that

Reply:
Updated the READMEs